### ISSUE TITLE
Decoder[BigDecimal] should not lose precision

### ISSUE THREAD

**Issue by gabriel-bezerra** [circe/circe#1101]
Some use cases of JSON are sensitive to the precision of the numbers in the JSON format: http://hl7.org/fhir/R4/json.html#primitive

Currently in circe 0.11.1 one can:

```scala
// using ammonite:
import $ivy.`io.circe::circe-parser:0.11.1`
```

- parse a number keeping precision:

```scala
val parsed = io.circe.parser.parse("0.000")
// parsed: Either[ParsingFailure, Json] = Right(JNumber(JsonDecimal("0.000")))
```

- encode a `BigDecimal` keeping precision:

```scala
val encoded = io.circe.Encoder[BigDecimal].apply(BigDecimal("0.000"))
// encoded: Json = JNumber(JsonBigDecimal(0.000))

val encodedString = encoded.noSpaces
// encodedString: String = "0.000"
```

- but not decode the parsed number keeping precision:

```scala
val decoded = parsed.map(io.circe.Decoder[BigDecimal].decodeJson)
// decoded: Either[ParsingFailure, Decoder.Result[BigDecimal]] = Right(Right(0))
```

The same holds for `java.math.BigDecimal`:

```scala
val encodedJava = io.circe.Encoder[java.math.BigDecimal].apply(new java.math.BigDecimal("0.000"))
// encodedJava: Json = JNumber(JsonBigDecimal(0.000))

val decodedJava = parsed.map(io.circe.Decoder[java.math.BigDecimal].decodeJson)
// decodedJava: Either[ParsingFailure, Decoder.Result[java.math.BigDecimal]] = Right(Right(0))
```

---

In general, it would be nice that, for any `s: String` abiding by the `number` grammar rule in  https://www.json.org whose decimal value can be represented by

> an arbitrary precision integer _unscaled value_ and a 32-bit integer _scale_ [...] (`unscaledValue` × 10^(-`scale`)).

(https://docs.oracle.com/javase/7/docs/api/java/math/BigDecimal.html)

one could have

```scala
def checkJava(s: String) = for {
    parsed <- io.circe.parser.parse(s)
    decoded <- io.circe.Decoder[java.math.BigDecimal].decodeJson(parsed)
    encoded = io.circe.Encoder[java.math.BigDecimal].apply(decoded)
  } yield encoded.noSpaces == s

def checkScala(s: String) = for {
    parsed <- io.circe.parser.parse(s)
    decoded <- io.circe.Decoder[scala.math.BigDecimal].decodeJson(parsed)
    encoded = io.circe.Encoder[scala.math.BigDecimal].apply(decoded)
  } yield encoded.noSpaces == s
```

holding `Right(true)`.

```scala
checkJava("0.1") //: Either[Error, Boolean] = Right(true)
checkJava("0.01") //: Either[Error, Boolean] = Right(true)
checkJava("0.010") //: Either[Error, Boolean] = Right(false) -- should be Right(true) instead

checkScala("0.1") //: Either[Error, Boolean] = Right(true)
checkScala("0.01") //: Either[Error, Boolean] = Right(true)
checkScala("0.010") //: Either[Error, Boolean] = Right(false) -- should be Right(true) instead
```

UPDATE:

More generally, given that the  encoding of `java.math.BigDecimal` seems correct and that its `equals` takes the `scale` into account:

```scala
def checkJava(s: String) = for {
    parsed <- io.circe.parser.parse(s)
    decoded <- io.circe.Decoder[java.math.BigDecimal].decodeJson(parsed)
    encoded = io.circe.Encoder[java.math.BigDecimal].apply(decoded)
  } yield new java.math.BigDecimal(s) == new java.math.BigDecimal(encoded.noSpaces)

checkJava("2") //: Either[Error, Boolean] = Right(true)
checkJava("2.0") //: Either[Error, Boolean] = Right(false) -- should be Right(true) instead
checkJava("2.01") //: Either[Error, Boolean] = Right(true)
```

---

Something to notice is that `equals` of `scala.math.BigDecimal` does not work in the same way of `java.math.BigDecimal` -- maybe it influences the result of the current test suite:

```scala
val scalaEquals = scala.math.BigDecimal("0.000") == scala.math.BigDecimal("0")
// scalaEquals: Boolean = true

val javaEquals0 = new java.math.BigDecimal("0.000") == new java.math.BigDecimal("0")
// javaEquals0: Boolean = false

val javaEquals1 = new java.math.BigDecimal("0.000") == new java.math.BigDecimal("0.000")
// javaEquals1: Boolean = true
```

**Comment by travisbrown** [comment]
Working on this now and it will be fixed in 0.12.0. Thanks for the thorough report, and sorry about the long delay in getting it fixed!

**Comment by plokhotnyuk** [comment]
@travisbrown great! feel free to peek and adopt [this](https://github.com/plokhotnyuk/jsoniter-scala/blob/db689583dc35cd3dd457e9edb0b7fbcc10a5d508/jsoniter-scala-core/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonReader.scala#L1614) routine which parse `BigDecimal` more efficiently for small values and with algorithmic complexity `O(n^1.5)` (instead of `O(n^2)` currently), where `n` is a number of digits in the mantissa

**Comment by travisbrown** [comment]
Fixed in #1217 and #1218.

**Comment by fedefernandez** [comment]
Hi @travisbrown!

I'm getting this behaviour with circe 0.13.0

```scala
  import io.circe._
  import io.circe.parser._
  import io.circe.generic.semiauto._
  val input = """{"bd": 0.00}"""
  case class Test(bd: BigDecimal)
  implicit val decoder: Decoder[Test] = deriveDecoder
  IO.fromEither(parse(input))
    .flatMap(json => IO.fromEither(decoder.decodeJson(json)))
    .map(_.bd.toString)
    .map(println(_))
    .unsafeRunSync()
```

```bash
```
